snuba/utils/streams/backends/local/storages/file.py

```diff
@@ -0,0 +1,133 @@
+import itertools
+import pickle
+from datetime import datetime
+from functools import cached_property
+from pathlib import Path
+from struct import Struct
+from typing import (
+    BinaryIO,
+    MutableMapping,
+    MutableSequence,
+    Optional,
+    Sequence,
+    Tuple,
+)
+
+from snuba.utils.codecs import Codec
+from snuba.utils.streams.backends.local.storages.abstract import (
+    MessageStorage,
+    PartitionDoesNotExist,
+    TopicDoesNotExist,
+    TopicExists,
+)
+from snuba.utils.streams.types import Message, Partition, Topic, TPayload
+
+
+class PickleCodec(Codec[bytes, Tuple[TPayload, datetime]]):
+    def encode(self, value: Tuple[TPayload, datetime]) -> bytes:
+        return pickle.dumps(value)
+
+    def decode(self, value: bytes) -> Tuple[TPayload, datetime]:
+        return pickle.loads(value)
+
+
+class FilePartition:
+    def __init__(self, path: Path) -> None:
+        self.__path = path
+
+    def exists(self) -> bool:
+        return self.__path.exists()
+
+    @cached_property
+    def reader(self) -> BinaryIO:
+        return self.__path.open("rb")
+
+    @cached_property
+    def writer(self) -> BinaryIO:
+        return self.__path.open("ab")
+
+
+class FileMessageStorage(MessageStorage[TPayload]):
+    def __init__(
+        self,
+        directory: str,
+        codec: Codec[bytes, Tuple[TPayload, datetime]] = PickleCodec(),
+    ) -> None:
+        self.__directory = Path(directory)
+        assert self.__directory.exists() and self.__directory.is_dir()
+        self.__codec = codec
+
+        self.__record_header = Struct("!L")
+        self.__topic_partition_cache: MutableMapping[
+            Topic, Sequence[FilePartition]
+        ] = {}
+
+    def create_topic(self, topic: Topic, partitions: int) -> None:
+        topic_path = self.__directory / topic.name
+        if topic_path.exists():
+            raise TopicExists(topic)
+
+        topic_path.mkdir()
```

Comment:
What about adding a method to cleanup the directory (remove all the topics, etc.)?  Since the directory is passed in by the caller, it may or may not be temporary. Then the caller may need to clean it up before moving on. Since this class is creating files and subdirectories inside the storage directory provided, I think it should provide a way to clean it up.

Reply:
I agree we should preferably use temp directories. What about making it the default behavior and making the path parameter optional? If there is no path provided this creates a temporary directory. Or could we need that directory outside this class ?

Reply:
Here is how that would look:

```diff
diff --git a/snuba/utils/streams/backends/local/storages/file.py b/snuba/utils/streams/backends/local/storages/file.py
index f409a789..f166cfa1 100644
--- a/snuba/utils/streams/backends/local/storages/file.py
+++ b/snuba/utils/streams/backends/local/storages/file.py
@@ -5,6 +5,7 @@ from datetime import datetime
 from functools import cached_property
 from pathlib import Path
 from struct import Struct
+from tempfile import TemporaryDirectory
 from typing import (
     BinaryIO,
     MutableMapping,
@@ -59,9 +60,16 @@ class InvalidChecksum(ValueError):
 class FileMessageStorage(MessageStorage[TPayload]):
     def __init__(
         self,
-        directory: str,
+        directory: Optional[str] = None,
         codec: Codec[bytes, Tuple[TPayload, datetime]] = PickleCodec(),
     ) -> None:
+        if directory is None:
+            # We need to maintain a reference to the ``TemporaryDirectory``
+            # even though we won't be using it directly since it (and its
+            # contents) will be deleted once the instance is garbage collected.
+            self.__temporary_directory = TemporaryDirectory()
+            directory = self.__temporary_directory.name
+
         self.__directory = Path(directory)
         assert self.__directory.exists() and self.__directory.is_dir()
         self.__codec = codec
diff --git a/tests/utils/streams/backends/test_local.py b/tests/utils/streams/backends/test_local.py
index 802c7560..faf32dc0 100644
--- a/tests/utils/streams/backends/test_local.py
+++ b/tests/utils/streams/backends/test_local.py
@@ -99,12 +99,8 @@ class LocalStreamsMemoryStorageTestCase(LocalStreamsTestMixin, TestCase):
 
 
 class LocalStreamsFileStorageTestCase(LocalStreamsTestMixin, TestCase):
-    def setUp(self) -> None:
-        self.directory = TemporaryDirectory()
-        super().setUp()
-
     def get_message_storage(self) -> MessageStorage[int]:
-        return FileMessageStorage(self.directory.name)
+        return FileMessageStorage()
 
     def test_unaligned_offset(self) -> None:
         topic = Topic(uuid.uuid1().hex)
```

Thinking about it more, I am a little hesitant to actually bring this in because I think that this use case (where messages are deleted once the storage is garbage collected) isn't one that this class should be targeting. If the goal is to have a message storage that is temporary, wouldn't the memory storage be a better fit than the file storage? I don't anticipate any situations where the contents of the temporary store are so large that we're exceeding the capacity of any reasonably provisioned machine, which is the only reason unrelated to persistence that I can think of where file storage would be preferable over memory. The only case that can think this of where this temporary directory default would be helpful is testing the file storage backend itself (which seems like a special case, almost by by definition.)

Reply:
Makes sense to me. Then back to the original comment, I would add a method to either remove topics or just to wipe the path. The client does not know the structure of the directory they provide, so it feels they should not tweak with files and directory to clean up the storage.